core/trino-main/src/main/java/io/trino/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -188,22 +208,36 @@ public GroupedAccumulator createGroupedAccumulator()
             accumulator = instantiateGroupedAccumulator(inputChannels, maskChannel);
         }
 
-        if (orderByChannels.isEmpty()) {
+        if (!orderByChannels.isEmpty()) {
+            accumulator = new OrderingGroupedAccumulator(accumulator, sourceTypes, orderByChannels, orderings, pagesIndexFactory);
+        }
+
+        if (!spillEnabled || (!hasDistinct() && !hasOrderBy())) {
             return accumulator;
         }
 
-        return new OrderingGroupedAccumulator(accumulator, sourceTypes, orderByChannels, orderings, pagesIndexFactory);
+        verify(accumulator instanceof FinalOnlyGroupedAccumulator, "FinalOnlyGroupedAccumulator is not used for final distincting/ordering aggregations");
```

Comment:
The `ClassCastException` thrown usually gives a better message (as it also mentions the class of the object passed)